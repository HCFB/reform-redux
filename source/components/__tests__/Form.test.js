import { shallow, mount } from 'enzyme';
import { object } from 'prop-types';
import { setImmediate } from 'timers';
import { createElement, Component } from 'react';
import { Form, Field } from '../../index';

describe('components / Form', () => {
  beforeEach(() => {
    global.formContext = {
      context: { store: global.store },
    };
  });

  it('snapshot', () => {
    const snapshot = shallow(createElement(Form, { path: 'form' }, ''), global.formContext);
    expect(snapshot).toMatchSnapshot();
  });

  it('if component has not "path" prop then throw error', () => {
    expect(() => shallow(createElement(Form, {}), global.formContext)).toThrow(
      'The `path` prop is required.',
    );
  });

  it('child component will get right context', () => {
    expect.assertions(3);

    class CustomInput extends Component {
      static contextTypes = {
        _reformRedux: object,
      };

      componentDidMount() {
        expect(Object.keys(this.context._reformRedux)).toEqual(
          expect.arrayContaining(['form', 'field']),
        );

        expect(Object.keys(this.context._reformRedux.form)).toEqual(
          expect.arrayContaining([
            'name',
            'path',
            'fieldsCount',
            'registerField',
            'unregisterField',
            'resetForm',
            'updateForm',
          ]),
        );

        expect(Object.keys(this.context._reformRedux.field)).toEqual(
          expect.arrayContaining([
            'changeFieldsValues',
            'changeFieldValue',
            'setFieldErrors',
            'setFieldsErrors',
            'setFieldDisabled',
            'setFieldsDisabled',
            'resetField',
            'resetFields',
          ]),
        );
      }

      render() {
        return '';
      }
    }

    mount(
      createElement(
        Form,
        { path: 'form' },
        createElement(CustomInput, { name: 'test', component: 'input' }),
      ),
      global.formContext,
    );
  });

  it('form initialisation in the store after component mount', () => {
    mount(
      createElement(
        Form,
        { path: 'form' },
        createElement(Field, { name: 'test', component: 'input' }),
      ),
      global.formContext,
    );

    expect(global.store.getState().form).toEqual({
      fields: { test: { changed: false, disabled: false, errors: [], valid: true, value: '' } },
      submitted: false,
      submitting: false,
      changed: false,
      valid: true,
    });
  });

  it('form component has default prop onSubmit', () => {
    const component = shallow(createElement(Form, { path: 'form' }), global.formContext);

    expect(typeof component.find('form').prop('onSubmit')).toBe('function');
  });

  it('form component onSubmit behavior', done => {
    expect.assertions(4);

    const validate = value => {
      expect(global.store.getState().form.submitting).toBeTruthy();

      if (value.length < 2) {
        return 'Value must be at least 2 characters';
      }
    };
    const onSubmit = jest.fn();

    const component = mount(
      createElement(Form, { path: 'form', onSubmit }, [
        createElement(Field, { key: 0, name: 'test1', value: '12', component: 'input', validate }),
        createElement(Field, { key: 1, name: 'test2', value: '12', component: 'input', validate }),
      ]),
      global.formContext,
    );

    component.find('form').simulate('submit');

    setImmediate(() => {
      expect(global.store.getState().form.submitting).toBeFalsy();
      expect(onSubmit).toBeCalledWith(
        {
          test1: {
            disabled: false,
            errors: [],
            valid: true,
            changed: false,
            value: '12',
          },
          test2: {
            disabled: false,
            errors: [],
            valid: true,
            changed: false,
            value: '12',
          },
        },
        expect.anything(),
      );
      done();
    });
  });

  it('form component onSubmitFailed behavior', done => {
    const validate = value => {
      if (value.length < 2) {
        return 'Value must be at least 2 characters';
      }
    };
    const onSubmitFailed = jest.fn();

    const component = mount(
      createElement(Form, { path: 'form', onSubmitFailed }, [
        createElement(Field, { key: 0, name: 'test1', value: '1', component: 'input', validate }),
        createElement(Field, { key: 1, name: 'test2', value: '1', component: 'input', validate }),
      ]),
      global.formContext,
    );

    component.find('form').simulate('submit');

    setImmediate(() => {
      expect(onSubmitFailed).toBeCalledWith(
        {
          test1: {
            disabled: false,
            errors: ['Value must be at least 2 characters'],
            valid: false,
            changed: false,
            value: '1',
          },
          test2: {
            disabled: false,
            errors: ['Value must be at least 2 characters'],
            valid: false,
            changed: false,
            value: '1',
          },
        },
        {
          test1: {
            disabled: false,
            errors: ['Value must be at least 2 characters'],
            valid: false,
            changed: false,
            value: '1',
          },
          test2: {
            disabled: false,
            errors: ['Value must be at least 2 characters'],
            valid: false,
            changed: false,
            value: '1',
          },
        },
        expect.anything(),
      );
      done();
    });
  });
});
