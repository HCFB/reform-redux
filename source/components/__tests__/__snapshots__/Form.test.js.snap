// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components / Form snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Form
    onSubmit={[Function]}
    onSubmitFailed={[Function]}
    path="form"
  >
    
  </Form>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "",
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": "",
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "",
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": "",
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "context": Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
    },
  },
}
`;
